ui/app/components/clients/charts/vertical-bar-stacked.ts

```diff
@@ -0,0 +1,139 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: BUSL-1.1
+ */
+
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+import { BAR_WIDTH, numericalAxisLabel } from 'vault/utils/chart-helpers';
+import { formatNumber } from 'core/helpers/format-number';
+import { parseAPITimestamp } from 'core/utils/date-formatters';
+import { flatGroup } from 'd3-array';
+import type { MonthlyChartData } from 'vault/vault/charts/client-counts';
+import type { ClientTypes } from 'core/utils/client-count-utils';
+
+interface Args {
+  chartHeight?: number;
+  chartLegend: Legend[];
+  chartTitle: string;
+  data: MonthlyChartData[];
+}
+
+interface Legend {
+  key: ClientTypes;
+  label: string;
+}
+interface AggregatedDatum {
+  x: string;
+  y: number;
+  legendX: string;
+  legendY: string[];
+}
+
+interface Base {
+  timestamp: string;
+  clientType: string;
+}
+
+type KeyDataItem = Base & {
+  [key in ClientTypes]?: number | undefined;
+};
+
+/**
+ * @module VerticalBarStacked
+ * Renders a stacked bar chart of counts for different client types over time. The which client types to render
+ * is mapped from the "key" values from the @legend arg
+ *
+ * @example
+ * <Clients::Charts::VerticalBarStacked
+ * @chartTitle="Total monthly usage"
+ * @data={{this.byMonthActivityData}}
+ * @chartLegend={{this.legend}}
+ * @chartHeight={{250}}
+ * />
+ */
+export default class VerticalBarStacked extends Component<Args> {
+  barWidth = BAR_WIDTH;
+  @tracked activeDatum: AggregatedDatum | null = null;
+
+  get chartHeight() {
+    return this.args.chartHeight || 190;
+  }
+
+  get dataKeys(): ClientTypes[] {
+    return this.args.chartLegend.map((l: Legend) => l.key);
+  }
+
+  label(legendKey: string) {
+    return this.args.chartLegend.find((l: Legend) => l.key === legendKey).label;
+  }
+
+  get chartData() {
+    let dataset: [string, number | undefined, string, KeyDataItem[]][] = [];
+    // each datum needs to be its own object
+    for (const key of this.dataKeys) {
+      const keyData: KeyDataItem[] = this.args.data.map((d: MonthlyChartData) => ({
+        timestamp: d.timestamp,
```

Comment:
At the moment we don't have any charts that aren't monthly charts, so making this key dynamic felt like over-complicating this work. Though this was implemented with abstraction in mind so adding getters for the xKey will be fairly straightforward.